integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/common/BenchUtils.scala

```diff
@@ -246,6 +247,126 @@ object BenchUtils {
     }
   }
 
+  /**
+   * Perform a diff of the results collected from two DataFrames, allowing for differences in
+   * precision.
+   *
+   * This is only suitable for data sets that can fit in the driver's memory.
+
+   * @param df1 DataFrame to compare.
+   * @param df2 DataFrame to compare.
+   * @param ignoreOrdering Sort the data collected from the DataFrames before comparing them.
+   * @param epsilon Allow for differences in precision when comparing floating point values.
+   */
+  def compareResults(
+      df1: DataFrame,
+      df2: DataFrame,
+      ignoreOrdering: Boolean,
+      maxErrors: Int = 10,
+      epsilon: Double = 0.00001): Unit = {
+
+    val result1: Seq[Seq[Any]] = collectResults(df1, ignoreOrdering)
+    val result2: Seq[Seq[Any]] = collectResults(df2, ignoreOrdering)
+
+    if (result1.length == result2.length) {
+      var errors = 0
+      var i = 0;
+      while (i < result1.length && errors < maxErrors) {
+        val l = result1(i)
+        val r = result2(i)
+        if (!rowEqual(l, r, epsilon)) {
+          println(s"Row $i:\n${l.mkString(",")}\n${r.mkString(",")}\n")
+          errors += 1
+          if (errors == maxErrors) {
+            println(s"Aborting comparison after reaching maximum of $maxErrors errors")
+          }
+        }
+        i += 1
+      }
+
+      if (errors==0) {
+        println(s"Results match")
+      }
+
+    } else {
+      println(s"Row counts do not match: ${result1.length} != ${result2.length}")
+    }
+  }
+
+  private def collectResults(df: DataFrame, ignoreOrdering: Boolean): Seq[Seq[Any]] = {
```

Comment:
considering this is a benchmarking utility, should we also provide an iterator in case the collected results are too big?

Reply:
Spark's `DataFrame.collect()` loads the full result set into the driver and returns an `Array`. If (when) we need to handle comparisons of larger results I think we would need an alternate approach like converting the results to single files, downloading them, and using Scala code (without Spark) to perform the diff.

Reply:
It is a relatively minor change to be able to do this with an iterator.  We have that support in the python unit tests.  The only issue is that you cannot compare the size of the results ahead of time, because you don't know it yet.  But I agree until we hit a situation where we need it there is no point in doing it.  Also it is very slow to try and do a comparison like that single threaded.  It might be better to truncate floating point values and do a few anti-joins to see if there is something in the left that is not in the right, and vise versa.  This might not handle duplicate rows, so we might need something there too, but it would scale a lot better.

Reply:
I'd like to know more about the iterator approach. I did look at the `toLocalIterator` method but this seems to load one partition at a time so I don't think that would work when we need ordering across the result set. Could you point me to an example where we do this in the Python tests?

Reply:
It loads one partition of the result at a time, so it does preserve ordering.  It is just rather slow because we are doing the comparison single threaded.

Reply:
checkout mortgage_test.py

Reply:
I added the option to use an iterator and have been testing this out and it seems to work well.